public/app/features/explore/state/actions.ts

```diff
@@ -530,6 +533,124 @@ export const runQueries = (exploreId: ExploreId): ThunkResult<void> => {
+      Math.floor((queryResponse.series[0].fields[0].values.length - 1) / 2)
+    );
+
+    const oldestResponseEndTimestamp = queryResponse.series[queryResponse.series.length - 1].fields[0].values.get(
+      Math.floor((queryResponse.series[0].fields[0].values.length - 1) / 2)
+    );
+
+    const newRaw: RawTimeRange = {
+      from:
+        direction === QueryDirection.forward
+          ? processDateTime(newestResponseEndTimestamp)
+          : processDateTime(range.raw.from),
+      to:
+        direction === QueryDirection.backward
+          ? processDateTime(oldestResponseEndTimestamp)
+          : processDateTime(range.raw.to),
+    };
+
+    const newTimeRange: TimeRange = {
+      from: dateMath.parse(newRaw.from, false, undefined) ?? dateTime(),
+      to: dateMath.parse(newRaw.to, true, undefined) ?? dateTime(),
+      raw: newRaw,
+    };
+
+    if (!hasNonEmptyQuery(queries)) {
+      dispatch(clearQueriesAction({ exploreId }));
+      dispatch(stateSave()); // Remember to save to state and update location
+      return;
+    }
+
+    // Some datasource's query builders allow per-query interval limits,
+    // but we're using the datasource interval limit for now
+    const minInterval = datasourceInstance?.interval;
+
+    stopQueryState(querySubscription);
+
+    const datasourceId = datasourceInstance?.meta?.id;
+
+    const queryOptions: QueryOptions = {
+      minInterval: minInterval ?? '1s',
+      // maxDataPoints is used in:
+      // Loki - used for logs streaming for buffer size, with undefined it falls back to datasource config if it supports that.
+      // Elastic - limits the number of datapoints for the counts query and for logs it has hardcoded limit.
+      // Influx - used to correctly display logs in graph
+      maxDataPoints: mode === ExploreMode.Logs && datasourceId === 'loki' ? undefined : containerWidth,
+      liveStreaming: live,
+      showingGraph,
+      showingTable,
+      mode,
+    };
+
+    const transaction = buildQueryTransaction(queries, queryOptions, newTimeRange, !!scanning);
+
+    let firstResponse = true;
+    dispatch(changeLoadingStateAction({ exploreId, loadingState: LoadingState.Loading }));
+
+    const newQuerySub = runRequest(datasourceInstance, transaction.request)
+      .pipe(
+        // Simple throttle for live tailing, in case of > 1000 rows per interval we spend about 200ms on processing and
+        // rendering. In case this is optimized this can be tweaked, but also it should be only as fast as user
+        // actually can see what is happening.
+        live ? throttleTime(500) : identity,
+        map((data: PanelData) => preProcessAppendPanelData(data, queryResponse, direction))
+      )
+      .subscribe((data: PanelData) => {
+        if (!data.error && firstResponse) {
+          // We save queries to the URL here so that only successfully run queries change the URL.
+          dispatch(stateSave());
```

Comment:
Are we saving the url here? I would assume we do not change anything that goes to url here anyway.

Reply:
Nope; that was just a thing from copypasted runQueries that was modified to get the relevant functionalities working. I'm gonna clean it up.